src/hotspot/cpu/x86/macroAssembler_x86_poly.cpp

```diff
@@ -0,0 +1,874 @@
+
+  // Cleanup
+  vpxorq(xmm0, xmm0, xmm0, Assembler::AVX_512bit);
+  vpxorq(xmm1, xmm1, xmm1, Assembler::AVX_512bit);
+  vpxorq(T0, T0, T0, Assembler::AVX_512bit);
+  vpxorq(T1, T1, T1, Assembler::AVX_512bit);
+  vpxorq(T2, T2, T2, Assembler::AVX_512bit);
+  vpxorq(C0, C0, C0, Assembler::AVX_512bit);
+  vpxorq(C1, C1, C1, Assembler::AVX_512bit);
+  vpxorq(C2, C2, C2, Assembler::AVX_512bit);
+  vpxorq(A0, A0, A0, Assembler::AVX_512bit);
+  vpxorq(A1, A1, A1, Assembler::AVX_512bit);
+  vpxorq(A2, A2, A2, Assembler::AVX_512bit);
+  vpxorq(A3, A3, A3, Assembler::AVX_512bit);
+  vpxorq(A4, A4, A4, Assembler::AVX_512bit);
+  vpxorq(A5, A5, A5, Assembler::AVX_512bit);
+  vpxorq(B0, B0, B0, Assembler::AVX_512bit);
+  vpxorq(B1, B1, B1, Assembler::AVX_512bit);
+  vpxorq(B2, B2, B2, Assembler::AVX_512bit);
+  vpxorq(B3, B3, B3, Assembler::AVX_512bit);
+  vpxorq(B4, B4, B4, Assembler::AVX_512bit);
+  vpxorq(B5, B5, B5, Assembler::AVX_512bit);
+  vpxorq(R0, R0, R0, Assembler::AVX_512bit);
+  vpxorq(R1, R1, R1, Assembler::AVX_512bit);
+  vpxorq(R2, R2, R2, Assembler::AVX_512bit);
+  vpxorq(R1P, R1P, R1P, Assembler::AVX_512bit);
+  vpxorq(R2P, R2P, R2P, Assembler::AVX_512bit);
+  evmovdquq(A0, Address(rsp, 64*3), Assembler::AVX_512bit);
+  evmovdquq(A0, Address(rsp, 64*4), Assembler::AVX_512bit);
+  evmovdquq(A0, Address(rsp, 64*5), Assembler::AVX_512bit);
+  evmovdquq(A0, Address(rsp, 64*0), Assembler::AVX_512bit);
+  evmovdquq(A0, Address(rsp, 64*1), Assembler::AVX_512bit);
+  evmovdquq(A0, Address(rsp, 64*2), Assembler::AVX_512bit);
+  addq(rsp, 512/8*6); // (powers of R)
+}
+
+// This function consumes as many whole 16-byte blocks as available in input
+// After execution, input and length will point at remaining (unprocessed) data
+// and accumulator will point to the current accumulator value
+//
+void MacroAssembler::poly1305_process_blocks(Register input, Register length, Register accumulator, Register R)
+{
+  // Register Map:
+  //     input        = rdi;
+  //     length       = rbx;
+  //     accumulator  = rcx;
+  //     R            = r8;
+
+  const Register a0 = rsi;  // [in/out] accumulator bits 63..0
+  const Register a1 = r9;   // [in/out] accumulator bits 127..64
+  const Register a2 = r10;  // [in/out] accumulator bits 195..128
+  const Register r0 = r11;  // R constant bits 63..0
+  const Register r1 = r12;  // R constant bits 127..64
+  const Register c1 = r8;   // 5*R (upper limb only)
+
+  Label L_process16Loop, L_process16LoopDone;
+
+  // Load R
+  movq(r0, Address(R, 0));
+  movq(r1, Address(R, 8));
+
+  // Compute 5*R (Upper limb only)
+  movq(c1, r1);
+  shrq(c1, 2);
+  addq(c1, r1); // c1 = r1 + (r1 >> 2)
+
+  // Load accumulator
+  movq(a0, Address(accumulator, 0));
+  movq(a1, Address(accumulator, 8));
+  movzbq(a2, Address(accumulator, 16));
+
+  // VECTOR LOOP: Minimum of 256 bytes to run vectorized code
+  cmpl(length, 16*16);
+  jcc(Assembler::less, L_process16Loop);
+
+  poly1305_process_blocks_avx512(input, length,
```

Comment:
Since entire code is based on 512 bit encoding misalignment penalty may be costly here. A scalar peel handling (as done in tail) for input portion before a 64 byte aligned address  could further improve the performance for large block sizes.

Reply:
Hmm.. interesting. Is this for loading? `evmovdquq` vs `evmovdqaq`? I was actually looking at using evmovdqaq but there is no encoding for it yet (And just looking now on uops.info, they seem to have identical timings? perhaps their measurements are off..). There are quite a few optimizations I tried (and removed) here, but not this one..

Perhaps to have a record, while its relatively fresh in my mind.. since there is a 8-block (I deleted a 16-block vector multiply), one can have a peeled off version for just 256 as the minimum payload.. In that case we only need R^1..R^8, (not R^1..R^16). I also tried loop stride of 8 blocks instead of 16, but that gets quite bit slower (20ish%?).. There was also a version that did a much better interleaving of multiplication and loading of next message block into limbs.. There is potentially a better way to 'devolve' the vector loop at tail; ie. when 15-blocks are left, just do one more 8-block multiply, all the constants are already available..

I removed all of those eventually. Even then, the assembler code currently is already fairly complex. The extra pre-, post-processing and if cases, I was struggling to keep up myself. Maybe code cleanup would have helped, so it _is_ possible to bring some of that back in for extra 10+%? (There is a branch on my fork with that code)

I guess that's my long way of saying 'I don't want to complicate the assembler loop'?

Reply:
> And just looking now on uops.info, they seem to have identical timings?

Actual instruction being used (aligned vs unaligned versions) doesn't matter much here, because it's a dynamic property of the address being accessed: misaligned accesses that cross cache line boundary incur a penalty. Since the cache line size is 64 byte in size, every misaligned 512-bit access is penalized.

Reply:
I collected performance counters for the benchmark included with the patch and its showing around 30% of 64 byte loads were spanning across the cache line.

 Performance counter stats for 'java -jar target/benchmarks.jar -f 1 -wi 1 -i 2 -w 30 -p dataSize=8192':

      122385646614      cycles                                                      
      328096538160      instructions              #    2.68  insn per cycle         
       64530343063      MEM_INST_RETIRED.ALL_LOADS                                   
       22900705491      MEM_INST_RETIRED.ALL_STORES                                   
       19815558484      MEM_INST_RETIRED.SPLIT_LOADS                                   
         701176106      MEM_INST_RETIRED.SPLIT_STORES    

Presence of scalar peel loop before the vector loop can save this penalty. 
We should also extend the scope of optimization (preferably in this PR or in subsequent one) to optimize [MAC computation routine accepting ByteBuffer.](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/com/sun/crypto/provider/Poly1305.java#L116),

Reply:
To close this thread.. @jatin-bhateja and I talked and realized that it is not possible to re-align input here. At least not with peeling with scalar loop. Scalar loop peels full blocks only (i.e. 16 bytes at a time). So out of 64 positions, 1 is already aligned, 3 could be aligned with the right peel, and 60 will land badly regardless.